Support for project JDK settings.

sdedic opened the pull request:
This PR is based on prototype done by @lahodaj . I have just updated it to fit into the current codebase + integrated with the new JDK configuration command.

- NBLS now has a new setting: `netbeans.project.jdkhome`. This holds a path to a JDK that is used as a **default platform**.
- to avoid possible issues with default platform changing (i.e. changes from JDK21 to JDK8, ouch!), NBLS is still restarted when `netbeans.project.jdhome` changes

The **Project Runtime Java** setting in **JDK Configuration** command was changed, so it configures **both** JDK table in `java.configuration.runtimes` (JDT) and `netbeans.project.jdkhome` (NBLS).

![jdk-config](https://github.com/apache/netbeans/assets/26788611/8e5b376f-69b5-4a67-b100-8f91d0bf53c0)
// cc: @jisedlac -- pls review, can't put you as a reviewer.

jisedlac commented:
Currently the feature is designed to provide one item/checkbox per setting. It displays the setting key and the current value. This way we can use the first step of the "wizard" as an immediate overview of the current settings.

If I understand correctly, you want to merge two different settings for the Java Runtime for Projects - the current for RH and a new one for NBLS. How will be the other setting key and its value displayed in the first step? It's not clear from the provided screenshot.

I understand the motivation to join the related settings into one item, but that's not the way it's designed right now. Note that the Maven Runtime JAVA_HOME setting also controls the JDK used for running Maven projects. Using the suggested approach, you should probably merge this one, too.

I'm not ready to agree with the current solution. We should either discuss in more details to make it consistent with all the other settings, or start with a separate item/checkbox if we need to deliver something asap.

sdedic commented:
OK, to summarize offline discussion:
- currently the feature is designed as 'UI to change settings'. Following this course, the setting for NBLS should be separate item in the list.

Looking at the overall user experience:
- when RH+NBLS are both active (as it is for example for oci-devops extension users), settings of those should be consistent.
- since the user can work independently in the vscode Terminal window, or execute maven commands from Projects view (Maven for Java extension), even integrated terminal settings and Maven terminal settings should be somehow consistent with the `java.configuration.runtimes` and `netbeans.project.jdkhome` so Maven runs the same way

The current PR implementation mixes those two approaches (does not separate settings, and does not ensure complete consistency).

Let's discuss what path should be followed.

lahodaj commented on java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/project/JavaPlatformProviderOverride.java:
```diff
@@ -0,0 +1,124 @@
+import org.openide.util.Lookup;
+import org.openide.util.lookup.ServiceProvider;
+
+@ServiceProvider(service=JavaPlatformProvider.class, position=10_000)
```
I realized just now (sorry) - maybe this should be an abstract class, with an implementation with `@ServiceProvider` in `nbcode`? So that ordinary IDE does not go through this code (it should, hopefully, be a noop, but still.)

sdedic replied:
Addressed in  2a1738f

lahodaj reviewed (commented):
Overall looks (obviously) good to me, with one minor comment in JavaPlatformProviderOverride and I not an expert on `settings.ts`.

lahodaj commented:
Thanks for doing this, BTW!

sdedic commented:
I reworked the PR as discussed in the conversation. Now there are separate settings for project JDKs for RH and NBLS:
![project-jdk](https://github.com/apache/netbeans/assets/26788611/a905682a-fbfc-4bad-ab76-b8d5c1c64b95)

I updated the NBLS runtime property to `machine-overridable` and allowed its definition in workspace scope.

dbalek reviewed (approved):
Looks fine

lahodaj commented on java/java.lsp.server/vscode/src/jdk/settings.ts:
```diff
@@ -306,7 +306,7 @@ class ProjectJavaSettings extends Setting {
                         path: jdk.javaHome
                     });
-                }
+                    }
```
Nit: this seems like a leftover change that should be undone.

lahodaj reviewed (approved):
Looks good to me. Thanks!

sdedic commented:
Rebased on latest master, retrofitted to handle #7555 - if projectJdk is defined, it's compared against project settings. If undefined, the NBLS runtime jdk is compared.

sdedic commented:
@petrovic-d pls review the changes in JDK validation.

petrovic-d commented:
Nitpick: I would rename [this method](https://github.com/apache/netbeans/blob/1c980f710ff56c7134627f12a0e822297f7c70bd/java/java.lsp.server/vscode/src/jdk/validation/validation.ts#L30) parameters to give it a bit more context. Maybe `languageServerJavaPath` and `projectJavaPath`, but it's up to you. 
Other than that it looks good to me.
